lua_modules/websocket/websocket.lua

```diff
@@ -0,0 +1,496 @@
+  end)
+  sk:on('reconnection', function(errcode)
+    print('Websocket reconnection. err:', errcode)
+  end)
+  sk:on("connection", function(sck, c)
+    -- print("Websocket got connection. Sending TCP msg. body:")
+    print(body)
+    -- m.sock = sck 
+    sck:send(body)
+    
+    m.pingStart()
+  end)
+  
+  m.sock = sk
+  sk:connect(port, host)
+  
+end
+
+-- currently only supports minimal frame decode
+m.lenExpected = 0
+m.buffer = ""
+-- private method
+function m.decodeFrame(frame)
+  
+  local data, fin, opcode
+  
+  if m.lenExpected > 0 then 
+    -- we should just raw append
+    data = frame 
+    -- print("Raw len:", string.len(data))
+    -- m.lenExpected = m.lenExpected - string.len(data)
+    
+    -- append the new data to the previous data
+    m.buffer = m.buffer .. data
+
+  else
+    -- we need to decode the headr from the frame 
+  
+    -- get FIN. 1 means msg is complete. 0 means multi-part
+    fin = string.byte(frame, 1)
+    fin = bit.isset(fin, 7)
+    -- print("FIN:", fin)
+    
+    -- get opcode
+    opcode = string.byte(frame, 1)
+    opcode = bit.clear(opcode, 4, 5, 6, 7) -- clear FIN and RSV
+    -- print("Opcode:", opcode)
+    
+
+    -- get 2nd byte as it has the payload length
+    -- msb of byte is mask, remaining 7 bytes is len 
+    local plen = string.byte(frame, 2)
+    local mask = bit.isset(plen, 7)
+    plen = bit.clear(plen, 7) -- remove the mask from the length
+    -- print("Frame len:", plen, " Mask:", mask)
+    m.lenExpected = plen
+    
+    if mask then 
+      -- print("We should not get a mask from server. Error.")
+      return
+    end
+    
+    data = string.sub(frame, 3) -- remove first 2 bytes, i.e. start at 3rd char
+  
+    if plen == 126 then 
+      -- read next 2 bytes
+      local extlen = string.byte(frame, 3)
+      local extlen2 = string.byte(frame, 4)
+      -- bit shift by one byte extlen 
+      extlen = bit.lshift(extlen, 8)
+      local actualLen = extlen + extlen2
+      -- print("ActualLen:", actualLen)
+      m.lenExpected = actualLen
+      
+      data = string.sub(data, 3) -- remove first 2 bytes
+      
+    elseif plen == 127 then 
+      -- print("Websocket lib does not support longer payloads yet")
+      -- return
+      data = string.sub(data, 5) -- remove first 4 bytes
+    end
+    
+    -- set the buffer to the current data since it's new
+    m.buffer = data
+  end 
+  
+  -- print("Our own count of len:", string.len(data))
+  
+  -- calc m.lenExpected for next time back into this method
+  m.lenExpected = m.lenExpected - string.len(data)
+  -- print("m.lenExpected:", m.lenExpected)
+  
+  -- we need to see if next time into decodeFrame we are just 
+  -- expecting more raw data without a frame header, i.e. happens if 
+  -- the TCP packet is > 1024
+  if m.lenExpected > 0 then
+    print("Websocket expecting " .. m.lenExpected .. " more chars of data")
+  else 
+    -- done with data. do callback 
+    m.lenExpected = 0 
```

Comment:
This should only be necessary in the case mentioned above

Reply:
In practice, what I've seen is you get lots of "receive" events from the TCP connection on the incoming data and you don't get all the data, however, you know how much you're going to get due to the frame header. So I'm only trying to solve for the multiple function calls that occur until you get the whole amount of data. Then I give you the callback. I don't think the scenario of getting more data for the next frame is part of the standard. So I'm pretty sure I'm safe here.

Reply:
I think this has nothing to do with the standard. Its only about packaging the information in one or more packages.
Still might be safe in most cases. Maybe a warning would be nice at first, so in case it happens the user can spot it esily.
But I am not the one to decide what goes in anyways.

Reply:
I think you're thinking of m.lenExpected in a different way than how it's being used. But, it's my fault for printing this message out. I just commented out that line since it really was just a debug line during development so I could follow what was happening. Check out the new commit. I think this cleans this part up.